### PR TITLE
Optimize distributed solve match fetch

### PR DESCRIPTION
**Pull request by trautmane**
by adding resolvedTilesWithMatchesFrom API endpoint and integrating it into AffineAlignBlockWorker.  

This is just a first draft that produces equivalent results to the original distributed solve implementation.  Further refinement is expected.  The key optimization achieved was reduction in the number of web service requests for each solve and reduction in the amount of data returned for each request.  For a distributed solve of hess_wafer_53::c000_s095_v01, the total number of service GET requests were reduced from 5959 to 208.

### CODE REVIEW

File: render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java [hunk 1]
```diff
@@ -30,37 +35,101 @@ public ResolvedTileSpecsWithMatchPairs(final ResolvedTileSpecCollection resolved
...
+        final Integer numberOfTilesToKeep = tileIdsToKeep == null ? null : tileIdsToKeep.size();
+
+        LOG.info("normalize: entry, process {} tiles and {} pairs with {} tileIdsToKeep and maxZDistance {}",
+                 resolvedTileSpecs.getTileCount(), matchPairs.size(), numberOfTilesToKeep, maxZDistance);
+
+        if ((maxZDistance != null) && (maxZDistance <= 0)) {
+            throw new IllegalArgumentException("maxZDistance must be >= 0 or null");
+        }
+
+        // remove unwanted tiles (which will later cause associated math pairs to be removed)
```
Comment (minnerbe) [inline render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java:83]: typo `math -> match`

File: render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java [hunk 2]
```diff
@@ -30,37 +35,101 @@ public ResolvedTileSpecsWithMatchPairs(final ResolvedTileSpecCollection resolved
...
+        if (tileIdsToKeep != null) {
+
+            // track and log removal info if debugging
+            final boolean isDebugEnabled = LOG.isDebugEnabled();
+            final Set<String> beforeTileIds = isDebugEnabled ? new HashSet<>(resolvedTileSpecs.getTileIds()) : null;
+
+            resolvedTileSpecs.removeDifferentTileSpecs(tileIdsToKeep);
```
Comment (minnerbe) [inline render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java:90]: The 'standard name' for this operation in the `Collection` interface would be `retainAll()`.

File: render-app/src/main/java/org/janelia/alignment/spec/TileSpec.java [hunk 3]
```diff
@@ -118,6 +118,10 @@ public Double getZ() {
         return z;
     }
 
+    public Integer getIntegralZ() {
```
Comment (minnerbe) [inline render-app/src/main/java/org/janelia/alignment/spec/TileSpec.java:121]: I would probably name this `getIntegerZ()` to remove ambiguities with the mathematical operation of integration.

File: render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java [hunk 4]
```diff
@@ -30,37 +35,101 @@ public ResolvedTileSpecsWithMatchPairs(final ResolvedTileSpecCollection resolved
...
+        final List<CanvasMatches> normalizedMatchPairs = new ArrayList<>(matchPairs.size());
+
+        for (final CanvasMatches pair : matchPairs) {
+            final TileSpec pTileSpec = resolvedTileSpecs.getTileSpec(pair.getpId());
+            final TileSpec qTileSpec = resolvedTileSpecs.getTileSpec(pair.getqId());
+            if ((pTileSpec != null) && (qTileSpec != null)) {
+                if (maxZDistance == null) {
                     normalizedMatchPairs.add(pair);
+                } else {
+                    final int zDistance = (int) Math.abs(pTileSpec.getZ() - qTileSpec.getZ());
```
Comment (minnerbe) [inline render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java:117]: Could this use the new `getIntegralZ()` method?

**Comment by minnerbe** [comment]
Other than the minor comments above, seems good to go. The reduction of the queries by an order of magnitude is great!

Thanks for cleaning up the `AffineAlignBlockWorker` a bit. I'm currently also cleaning this up while trying to split input and output, so I might have to rebase my branch, but this shouldn't be a big deal.

File: render-app/src/main/java/org/janelia/alignment/spec/TileSpec.java (hunk 3 above)
Comment (trautmane) [reply render-app/src/main/java/org/janelia/alignment/spec/TileSpec.java]: good point - done

File: render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java (hunk 2 above)
Comment (trautmane) [reply render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java]: Yeah - I did not realize that years ago when I wrote the original method - doh.  I've updated the name.

File: render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java (hunk 1 above)
Comment (trautmane) [reply render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java]: done

File: render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java (hunk 4 above)
Comment (trautmane) [reply render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java]: yes - actually decided to take it one step further and add a zDistanceFrom method to TileSpec instead

**Pull request merged**